src/node_file.cc

```diff
@@ -969,14 +970,78 @@ static void InternalModuleReadJSON(const FunctionCallbackInfo<Value>& args) {
   Local<Value> return_value[] = {
     String::NewFromUtf8(isolate,
                         &chars[start],
-                        v8::NewStringType::kNormal,
+                        NewStringType::kNormal,
                         size).ToLocalChecked(),
     Boolean::New(isolate, p < pe ? true : false)
   };
   args.GetReturnValue().Set(
     Array::New(isolate, return_value, arraysize(return_value)));
 }
 
+Isolate* _isolate = nullptr;
+std::vector<Local<Value>>* _exports = nullptr;
+std::vector<Local<Value>>* _reexports = nullptr;
+
+void addExport(const uint16_t* start, const uint16_t* end) {
+  MaybeLocal<String> specifier = String::NewFromTwoByte(_isolate, start,
+                                                        NewStringType::kNormal,
+                                                        end - start);
+  CHECK(!specifier.IsEmpty());
+  _exports->push_back(specifier.ToLocalChecked());
+}
+
+void addReexport(const uint16_t* start, const uint16_t* end) {
+  MaybeLocal<String> specifier = String::NewFromTwoByte(_isolate, start,
+                                                        NewStringType::kNormal,
+                                                        end - start);
+  CHECK(!specifier.IsEmpty());
+  _reexports->push_back(specifier.ToLocalChecked());
+}
+
+// Applies lexical analysis to detect named exports for transpiled modules only.
+static void InternalModuleParseCJS(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+
+  CHECK(args[0]->IsString());
+  Local<String> sourceString = args[0].As<String>();
+
+  unsigned int len = sourceString->Length();
+
+  uint16_t* source =
+    reinterpret_cast<uint16_t*>(malloc((len + 1) * sizeof(uint16_t)));
+
+  std::vector<Local<Value>> exports;
+  std::vector<Local<Value>> reexports;
+
+  _isolate = isolate;
+  _exports = &exports;
+  _reexports = &reexports;
+
+  // Write the UTF-16 String into a temporary buffer.
+  sourceString->Write(isolate, source);
+
+  bool ok = parseCJS(source, len, &addExport, &addReexport);
+
+  free(source);
+
+  _isolate = nullptr;
+  _exports = nullptr;
+  _reexports = nullptr;
+
+  if (!ok) {
+    Local<Value> return_value[] = { Array::New(isolate), Array::New(isolate) };
```

Comment:
Likely better just to return these as `Undefined` and create the arrays in javascript if necessary.